Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,234 @@
+Other eventing components in CoreCLR must already poll LTTNG for changes in the eventing state, so this design can utilize
+that same poll to inform the GC of changes.
+
+## Implementation Concerns
+
+An implementation of this spec must take care not to perturb the existing GC code base too much. The GC fires events through
+the use of macros generated by the ETW message compiler and carefully mocked by code generation for the other platform
+logging implementations. The eventing scheme in this document will need to provide implementations for all eventing
+macros used by the GC ([1]).
+
+These headers are often auto-generated. We will need to take care to re-use existing code generators if possible - after all,
+we do want it to be easy to add new events. It will likely be difficult to balance auto-generated code with the need for
+subtle custom modifications to event dispatch code.
+
+Tools (specifically PerfView) will want to be enlightened about dynamic events fired by the runtime. The extent to which
+we want to make this experience nice is up to us, but we will most likely want to ship PerfView support for any dynamic event
+that we end up shipping with the CLR GC.
+
+[1]: https://github.com/dotnet/coreclr/blob/cab0db6345a7941f75d991281bcc0079d28ba182/src/gc/env/etmdummy.h#L5-L57 
+
+### Concrete Example: Porting a single known event to IGCToCLREventSink
+
+The following steps illustrate what needs to be done to bring a single event over to `IGCToCLREventSink`:
+
+Two things are needed by the GC in order to fire an event: a way to determine if the event is on, and a way to fire
+the event. Events can be fired even if they aren't enabled (they occasionally are); the platform loggers will ignore the
+event if it is not enabled. However, the GC generally avoids doing expensive eventing-related operations if an event is
+not on.
+
+The GC generally uses the `ETW_EVENT_ENABLED` macro to query whether an event is on. `ETW_EVENT_ENABLED` will be
+implemented in terms of `GCEventStatus::IsEnabled` above, so you will need to define appropriate macros in order for your
+event to work here. This will likely mean that you will need to define a macro that turns the name of your event into a
+pair of a level and a keyword, which will determine whether or not your event is enabled.
+
+To fire the event, you can add your event's callback to `IGCToCLREventSink`:
+
+```c++
+class IGCToCLREventSink
+{
+    virtual void FireYourEvent(
+        /* your event arguments here... */
+    ) = 0;
+};
+```
+
+You can then define the `FireYourEvent` macro in `src/gc/env/etmdummy.h` to point to your new method on `IGCToCLREventSink`
+Your implementation of `FireYourEvent` in the EE will need to calculate any EE-specific data (e.g. `ClrInstanceId`) and
+then forward the event arguments onto the platform logger, which can be done with the `Fire` macros that the EE has access
+to (that we implemented for the GC).
```

Comment:
can you please also add an example of adding a dynamic event?

Reply:
will do!

Reply:
I added a short example in the next iteration, but it's hard to provide a full example without writing large amounts of boilerplate code. I plan on providing classes within the GC that greatly automate the adding of dynamic events, to the point where you should not have to write any binary serialization code for any new events that you add.